docs/src/content/en/docs/agents/processors.mdx

```diff
@@ -0,0 +1,216 @@
+  name: "moderated-agent",
+  instructions: "You are a helpful assistant",
+  model: openai("gpt-4o-mini"),
+  inputProcessors: [
+    new ModerationProcessor({
+      model: openai("gpt-4.1-nano"),
+      categories: ["hate", "harassment", "violence"],
+      threshold: 0.7,
+      strategy: "block",
+    }),
+  ],
+});
+```
+
+## Execution order
+
+Processors run in the order they appear in the array:
+
+```typescript
+inputProcessors: [
+  new UnicodeNormalizer(),
+  new PromptInjectionDetector(),
+  new ModerationProcessor(),
+];
+```
+
+For output processors, the order determines the sequence of transformations applied to the model's response.
+
+Memory processors (when memory is enabled) are automatically added to the agent's processor pipeline. See [Memory Processors](/docs/memory/memory-processors) for details on how memory processors integrate with manually configured processors.
+
+## Creating custom processors
+
+Custom processors implement the `InputProcessor` or `OutputProcessor` interface:
+
+### Custom input processor
+
+```typescript title="src/mastra/processors/custom-input.ts" showLineNumbers copy
+import type {
+  InputProcessor,
+  MastraMessageV2,
+  RuntimeContext,
+} from "@mastra/core";
+
+export class CustomInputProcessor implements InputProcessor {
+  name = "custom-input";
+
+  async processInput({
+    messages,
+    context,
+  }: {
+    messages: MastraMessageV2[];
+    context: RuntimeContext;
+  }): Promise<MastraMessageV2[]> {
+    // Transform messages before they reach the LLM
+    return messages.map((msg) => ({
+      ...msg,
+      content: {
+        ...msg.content,
+        content: msg.content.content.toLowerCase(),
+      },
+    }));
+  }
+}
+```
+
+### Custom output processor
+
+```typescript title="src/mastra/processors/custom-output.ts" showLineNumbers copy
+import type {
+  OutputProcessor,
+  MastraMessageV2,
+  RuntimeContext,
+} from "@mastra/core";
+
+export class CustomOutputProcessor implements OutputProcessor {
+  name = "custom-output";
+
+  async processOutputResult({
+    messages,
+    context,
+  }: {
+    messages: MastraMessageV2[];
+    context: RuntimeContext;
+  }): Promise<MastraMessageV2[]> {
+    // Transform messages after the LLM generates them
+    return messages.filter((msg) => msg.role !== "system");
+  }
+
+  async processOutputStream({
+    stream,
+    context,
+  }: {
+    stream: ReadableStream;
+    context: RuntimeContext;
+  }): Promise<ReadableStream> {
+    // Transform streaming responses
+    return stream;
+  }
+}
+```
+
+Both `processInput` and `processOutputResult` receive a `context` object containing execution metadata such as `threadId`, `resourceId`, and other runtime information.
+
+## Built-in Utility Processors
+
+Mastra provides utility processors for common tasks:
+
+**For security and validation processors**, see the [Guardrails](/docs/agents/guardrails) page for input/output guardrails and moderation processors.
```

Comment:
Ah I see yu mentioned it here, i would probably still add a small bullet above too